sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,343 @@
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var violations = FindViolations(c).ToList();
+                violations.ForEach(c.ReportDiagnostic);
+            },
+            SyntaxKind.MethodDeclaration);
+        }
+
+        public IEnumerable<Diagnostic> FindViolations(SyntaxNodeAnalysisContext context)
+        {
+            var methodSymbol = context.SemanticModel.GetDeclaredSymbol(context.Node) as IMethodSymbol;
+
+            if (methodSymbol == null ||
+                methodSymbol.Parameters.Length == 0 ||
+                methodSymbol.IsOverride)
+            {
+                return Enumerable.Empty<Diagnostic>();
+            }
+
+            var parametersToCheck = new Dictionary<string, ParameterData>();
+            foreach (var parameter in methodSymbol.Parameters.Where(IsTrackedParameter))
+            {
+                parametersToCheck[parameter.Name] = new ParameterData(parameter);
+            }
+
+            var parameterUsesInMethod = context
+                .Node
+                .DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .Where(id => parametersToCheck.Values.Any(p => p.MatchesIdentifier(id, context.SemanticModel)))
+                .ToArray();
+
+            foreach (var idName in parameterUsesInMethod)
+            {
+                var key = idName?.Identifier.ValueText ?? "";
+
+                ParameterData paramData;
+                if (!parametersToCheck.TryGetValue(key, out paramData) ||
+                    !paramData.ShouldReportOn)
+                {
+                    continue;
+                }
+
+                if (idName.Parent is EqualsValueClauseSyntax ||
+                    idName.Parent is AssignmentExpressionSyntax)
+                {
+                    paramData.ShouldReportOn = false;
+                    continue;
+                }
+
+                var symbolUsedAs = FindParameterUseAsType(idName, context.SemanticModel);
+                if (symbolUsedAs != null)
+                {
+                    paramData.AddUsage(symbolUsedAs);
+                }
+            }
+
+            return parametersToCheck
+                .Values
+                .Select(p => p.GetRuleViolation())
+                .WhereNotNull();
+        }
+
+        private static bool IsTrackedParameter(IParameterSymbol parameterSymbol)
+        {
+            var type = parameterSymbol.Type;
+            var isExcluded =
+                type.DerivesFrom(KnownType.System_Array) ||
+                type.IsValueType ||
+                type.Is(KnownType.System_String);
+
+            return !isExcluded;
+        }
+
+        private ITypeSymbol FindParameterUseAsType(IdentifierNameSyntax idRef, SemanticModel semanticModel)
+        {
+            if (idRef.Parent is ConditionalAccessExpressionSyntax)
+            {
+                var cond = idRef.Parent as ConditionalAccessExpressionSyntax;
```

Comment:
Don't use `as` cast when you are sure about the result it makes things harder to read.